src/common/TrackedOp.cc

```diff
@@ -286,41 +274,39 @@ bool OpTracker::register_inflight_op(TrackedOp *i)
   if (!tracking_enabled)
     return false;
 
-  RWLock::RLocker l(lock);
   uint64_t current_seq = ++seq;
-  uint32_t shard_index = current_seq % num_optracker_shards;
-  ShardedTrackingData* sdata = sharded_in_flight_list[shard_index];
-  assert(NULL != sdata);
+  uint32_t shard_index = current_seq & (sharded_in_flight_list.size() - 1);
+  ShardedTrackingData& sdata = sharded_in_flight_list[shard_index];
   {
-    Mutex::Locker locker(sdata->ops_in_flight_lock_sharded);
-    sdata->ops_in_flight_sharded.push_back(*i);
+    Mutex::Locker locker(sdata.ops_in_flight_lock_sharded);
+    sdata.ops_in_flight_sharded.push_back(*i);
     i->seq = current_seq;
   }
   return true;
 }
 
-void OpTracker::unregister_inflight_op(TrackedOp *i)
+void OpTracker::unregister_inflight_op(TrackedOpRef&& i)
 {
+  const auto ri = i.get();
   // caller checks;
-  assert(i->state);
+  assert(ri->state);
 
-  uint32_t shard_index = i->seq % num_optracker_shards;
-  ShardedTrackingData* sdata = sharded_in_flight_list[shard_index];
-  assert(NULL != sdata);
+  uint32_t shard_index = ri->seq & (sharded_in_flight_list.size() - 1);
```

Comment:
This is really ugly. Conceptually just a modulo but written in a way that exposes each horrid detail on a layer far away from the appropriate one.

Reply:
FIXED.